lib/block_view.go

```diff
@@ -3515,40 +3570,113 @@ func (bav *UtxoView) Preload(desoBlock *MsgDeSoBlock, blockHeight uint64) error
+func (bav *UtxoView) SetTransactorNonceEntry(nonceEntry *TransactorNonceEntry) {
+	if nonceEntry == nil {
+		glog.Errorf("SetTransactorNonceEntry: nil nonceEntry")
+		return
+	}
+
+	bav.TransactorNonceMapKeyToTransactorNonceEntry[nonceEntry.ToMapKey()] = nonceEntry
+}
+
+func (bav *UtxoView) DeleteTransactorNonceEntry(nonceEntry *TransactorNonceEntry) {
+	if nonceEntry == nil {
+		glog.Errorf("DeleteTransactorNonceEntry: nil nonceEntry")
 		return
 	}
-	bav.SetNextNonceForPKID(*pkidEntry.PKID, nextNonce)
+
+	nonceEntry.isDeleted = true
+	bav.SetTransactorNonceEntry(nonceEntry)
+}
+
+func (bav *UtxoView) GetTransactorNonceEntriesToDeleteAtBlockHeight(blockHeight uint64) []*TransactorNonceEntry {
+	dbExpiredNonceEntries, err := DbGetTransactorNonceEntriesToExpireAtBlockHeight(bav.Handle, blockHeight)
+	if err != nil {
+		glog.Errorf("GetTransactorNonceEntriesToDeleteAtBlockHeight: Error fetching expired nonce entries: %v", err)
+		return nil
+	}
+	// Add the db entries to the view
+	for _, dbNonceEntry := range dbExpiredNonceEntries {
+		// If the entry already exists, skip it and use what's in the view.
+		if _, exists := bav.TransactorNonceMapKeyToTransactorNonceEntry[dbNonceEntry.ToMapKey()]; exists {
+			continue
+		}
+		bav.SetTransactorNonceEntry(dbNonceEntry)
+	}
+	var transactorNoncesToExpire []*TransactorNonceEntry
+	for _, nonceEntry := range bav.TransactorNonceMapKeyToTransactorNonceEntry {
+		if nonceEntry.isDeleted {
+			continue
+		}
+		if nonceEntry.Nonce.ExpirationBlockHeight <= blockHeight {
+			transactorNoncesToExpire = append(transactorNoncesToExpire, nonceEntry)
+		}
+	}
+	return transactorNoncesToExpire
+}
+
+func (bav *UtxoView) ConstructNonceForPublicKey(publicKey []byte, blockHeight uint64) (*DeSoNonce, error) {
+	pkidEntry := bav.GetPKIDForPublicKey(publicKey)
+	if pkidEntry == nil || pkidEntry.isDeleted {
+		return nil, fmt.Errorf(
+			"ConstructNonceForPublicKey: No PKID entry found for public key %s",
+			PkToStringBoth(publicKey))
+	}
+	return bav.ConstructNonceForPKID(pkidEntry.PKID, blockHeight, 2)
 }
 
-func (bav *UtxoView) SetNextNonceForPKID(pkid PKID, nextNonce uint64) {
-	bav.PKIDToNextNonce[pkid] = nextNonce
+func (bav *UtxoView) ConstructNonceForPKID(pkid *PKID, blockHeight uint64, depth uint8) (*DeSoNonce, error) {
+	// construct nonce
+	expirationBuffer := InitialMaxNonceExpirationBlockHeightOffset
+	if bav.GlobalParamsEntry != nil && bav.GlobalParamsEntry.MaxNonceExpirationBlockHeightOffset != 0 {
+		expirationBuffer = bav.GlobalParamsEntry.MaxNonceExpirationBlockHeightOffset
+	}
+	nonce := DeSoNonce{
+		PartialID:             rand.Uint64(),
+		ExpirationBlockHeight: blockHeight + expirationBuffer,
+	}
+
+	// Make sure we don't have a collision.
+	nonceEntry, err := bav.GetTransactorNonceEntry(&nonce, pkid)
+	if err != nil {
+		return nil, errors.Wrapf(err, "ConstructNonceForPKID: ")
+	}
+	if nonceEntry == nil {
+		return &nonce, nil
+	}
+	if depth == 0 {
```

Comment:
Just for my own understanding, how do you plan on using the depth param?

Reply:
This really only serves to support transaction construction. In the event of a collision, we'll attempt to generate the nonce again.

It's always 2 on the initial call to this function. I'll add a check that depth <= 2. If we hit a collision three times when generating the nonce, we throw an error.